packages/core/src/service.ts

```diff
@@ -0,0 +1,122 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  BindingFilter,
+  Constructor,
+  ContextView,
+  inject,
+  InjectionMetadata,
+  MetadataInspector,
+  transformValueOrPromise,
+} from '@loopback/context';
+
+/**
+ * Metadata for `@service` injection
+ */
+export interface ServiceInjectionMetadata extends InjectionMetadata {
+  /**
+   * Do not check sub classes to match service bindings
+   */
+  skipSubClasses?: boolean;
+}
+
+/**
+ * `@service` injects a service with matching class
+ * @param serviceClass - Constructor for the service. If not provided, the value
+ * is inferred from the design:type of the parameter or property
+ *
+ * @example
+ * ```ts
+ * export class MyController {
+ *   constructor(@service(MyService) private myService: MyService) {}
+ *
+ *   @service()
+ *   private logger: Logger;
+ * }
+ * ```
+ */
+export function service(
+  serviceClass?: Constructor<unknown>,
+  metadata?: ServiceInjectionMetadata,
+) {
+  return inject(
+    '',
+    {decorator: '@service', ...metadata},
+    (ctx, injection, session) => {
+      let serviceType: Function | undefined = serviceClass;
+      if (!serviceType) {
+        if (typeof injection.methodDescriptorOrParameterIndex === 'number') {
+          serviceType = MetadataInspector.getDesignTypeForMethod(
+            injection.target,
+            injection.member!,
+          ).parameterTypes[injection.methodDescriptorOrParameterIndex];
+        } else {
+          serviceType = MetadataInspector.getDesignTypeForProperty(
+            injection.target,
+            injection.member!,
+          );
+        }
+      }
+      if (serviceType === Object || serviceType === Array) {
+        throw new Error('Service class cannot be inferred from design type');
```

Comment:
Please improve the error message to tell the user how to fix the problem. Should they provide the service class as the first decorator argument? Something else?